base/special/trig.jl

```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+                             7.70381505559019352791e-02)
+
+asin_p(t::Float32) = t*@horner(t,
+                             1.6666586697f-01,
+                             -4.2743422091f-02,
+                             -8.6563630030f-03)
+
+asin_q(t::Float32) = muladd(t, -7.0662963390f-01, 1.0f0)
+
+@inline function asin_kernel(t::Float64, x)
+    pio2_lo = 6.12323399573676603587e-17
+    s = sqrt_llvm(t)
+    p = asin_p(t) # numerator polynomial
+    q = asin_q(t) # denominator polynomial
+    if abs(x) >= 0.975 # |x| > 0.975
+        Rx = p/q
+        return flipsign(pi/2 - (2.0*(s + s*Rx) - pio2_lo), x)
+    else
+        s0 = reinterpret(Float64, (reinterpret(UInt64, s) >> 32) << 32)
+        c = (t - s0*s0)/(s + s0)
+        Rx = p/q
+        p = 2.0*s*Rx - (pio2_lo - 2.0*c)
+        q = pi/4 - 2.0*s0
+        return flipsign(pi/4 - (p-q), x)
+    end
+end
+@inline function asin_kernel(t::Float32, x)
+    s = sqrt_llvm(Float64(t))
+    p = asin_p(t) # numerator polynomial
+    q = asin_q(t) # denominator polynomial
+    Rx = p/q # rational approximation
+    flipsign(Float32(pi/2 - 2*(s + s*Rx)), x)
+end
+
+@noinline asin_domain_error(x) = throw(DomainError(x, "asin(x) is not defined for |x|>1."))
+function asin{T<:Union{Float32, Float64}}(x::T)
+    # Method :
+    #    Since  asin(x) = x + x^3/6 + x^5*3/40 + x^7*15/336 + ...
+    #    we approximate asin(x) on [0,0.5] by
+    #        asin(x) = x + x*x^2*R(x^2)
+    #    where
+    #        R(x^2) is a rational approximation of (asin(x)-x)/x^3
+    #    and its remez error is bounded by
+    #        |(asin(x)-x)/x^3 - R(x^2)| < 2^(-58.75)
+    #
+    #    For x in [0.5,1]
+    #        asin(x) = pi/2-2*asin(sqrt((1-x)/2))
+    #    Let y = (1-x), z = y/2, s := sqrt(z), and pio2_hi+pio2_lo=pi/2;
+    #    then for x>0.98
+    #        asin(x) = pi/2 - 2*(s+s*z*R(z))
+    #            = pio2_hi - (2*(s+s*z*R(z)) - pio2_lo)
+    #    For x<=0.98, let pio4_hi = pio2_hi/2, then
+    #        f = hi part of s;
+    #        c = sqrt(z) - f = (z-f*f)/(s+f)     ...f+c=sqrt(z)
+    #    and
+    #        asin(x) = pi/2 - 2*(s+s*z*R(z))
+    #            = pio4_hi+(pio4-2s)-(2s*z*R(z)-pio2_lo)
+    #            = pio4_hi+(pio4-2f)-(2s*z*R(z)-(pio2_lo+2c))
+    #
+    # Special cases:
+    #    if |x|>1, throw DomainError
+    absx = abs(x)
+    if absx >= one(T) # |x|>= 1
+        if absx == one(T)
+            return flipsign(T(pi)/2, x)
+        end
+        asin_domain_error(x)
+    elseif absx < one(T)/2
+        # if |x| sufficiently small, |x| is a good approximation
+        if absx < ASIN_X_MIN_THRESHOLD(T)
+            return x
+        end
+        # else if |x|<0.5 we use a rational approximation R(x)=p(x)/q(x) such that
+        # tan(x) ≈ x+x*R(x)
+        x² = x*x
+        Rx = asin_p(x²)/asin_q(x²) # rational approximation
+        return muladd(x, Rx, x)
+    end
+    # else 1/2 <= |x| < 1
+    t = (one(T) - absx)/2
+    return asin_kernel(t, x)
+end
```

Comment:
I don't know if "JuliaLang" prefers it to be here or in a separate file. I don't have an opinion myself, but I do get your point about moving it down.